Streetcode/Streetcode.WebApi/Utils/WebParsingUtils.cs

```diff
@@ -1,73 +1,341 @@
+            }
+
+            // OSM Nominatim does not like e.g "бульв. Івана  Світличного", use "Івана  Світличного бульвар" instead
+            else if (streetname.IndexOf("бульв.") != -1)
+            {
+                streetNameStringSearchOptimized = streetname.Substring(streetname.IndexOf(" ") + 1) + " бульвар";
+            }
 
-            if (!rows.Contains(newRow))
+            // OSM Nominatim does not like e.g "шосе Стратегічне", use "Стратегічне шосе" instead
+            else if (streetname.IndexOf("шосе ") != -1)
             {
-                rows.Add(newRow);
-                stringBuilder.AppendLine(newRow);
+                streetNameStringSearchOptimized = streetname.Substring(streetname.IndexOf(" ") + 1) + " шосе";
             }
+
+            // OSM Nominatim does not like e.g "містечко Військове", use "Військове містечко" instead
+            else if (streetname.IndexOf("містечко ") != -1)
+            {
+                streetNameStringSearchOptimized = streetname.Substring(streetname.IndexOf(" ") + 1) + " містечко";
+            }
+
+            // OSM Nominatim does not like e.g "в’їзд 2-й Криничний", use "2-й Криничний в’їзд" instead
+            else if (streetname.IndexOf("в’їзд") != -1)
+            {
+                streetNameStringSearchOptimized = streetname.Substring(streetname.IndexOf(" ") + 1) + " в’їзд";
+            }
+
+            return streetNameStringSearchOptimized;
         }
 
-        await File.WriteAllTextAsync(csvPath, stringBuilder.ToString(), Encoding.UTF8);
+        // Following method calls OSM Nominatim API for single address and returns lat and lon coordinates
+        // For some addresses nothing can be returned
+        public static async Task<Tuple<string, string>> CallAPICoords(string address)
+        {
+            try
+            {
+                WebClient webClient = new WebClient();
+                webClient.Headers.Add("user-agent", "Mozilla/4.0 (compatible; MSIE 6.0; Windows NT 5.2; .NET CLR 1.0.3705;)");
+                webClient.Headers.Add("Referer", "http://www.microsoft.com");
+                var jsonData = webClient.DownloadData("https://nominatim.openstreetmap.org/search?q=" + address + "&format=json&limit=1&addressdetails=1");
+                string jsonStr = Encoding.UTF8.GetString(jsonData);
+                Tuple<string, string> coords = ParseJsonToCoordinateTuple(jsonStr);
+                return coords;
+            }
+            catch (Exception ex)
+            {
+                Console.ForegroundColor = ConsoleColor.Red;
+                Console.WriteLine(ex.Message);
+                Console.ForegroundColor = ConsoleColor.White;
+            }
+
+            return null;
+        }
+
+        // Following method parses JSON from OSM Nominatim API and returns lat/lon tuple
+        public static Tuple<string, string> ParseJsonToCoordinateTuple(string json)
```

Comment:
This method could be converted to
`
    private static (string, string) ParseJsonToCoordinateTuple(string json)
    {
        var data = JsonConvert.DeserializeObject<List<Dictionary<string, object>>>(json);
        return (data?[0]["lat"].ToString() ?? string.Empty, data?[0]["lon"].ToString() ?? string.Empty);
    }
`

Reply:
I refactored this part of code but your version does not work